jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/SingleThreadedLogger.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.plugins.common;
+
+import com.google.common.util.concurrent.Futures;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.function.Consumer;
+
+/**
+ * Keeps all log messages in a sequential, deterministic order along with an additional footer that
+ * always appears below log messages.
+ */
+class SingleThreadedLogger {
+
+  /** ANSI escape sequence for moving the cursor up one line. */
+  private static final String CURSOR_UP_SEQUENCE = "\033[1A";
+
+  private final ExecutorService executorService = Executors.newSingleThreadExecutor();
+
+  private final Consumer<String> plainLogger;
+
+  private String footer = "";
+  private int footerLineCount = 0;
+
+  SingleThreadedLogger(Consumer<String> plainLogger) {
+    this.plainLogger = plainLogger;
+  }
+
+  /**
+   * Runs {@code messageLogger} asynchronously.
+   *
+   * @param messageLogger the {@link Runnable} intended to synchronously log a message to the
+   *     console
+   * @return a {@link Future} to track completion
+   */
+  public Future<Void> log(Runnable messageLogger) {
+    return log(messageLogger, footer, footerLineCount);
+  }
+
+  /**
+   * Sets the footer asynchronously.
+   *
+   * @param footer the footer
+   * @param lineCount the number of lines in the footer
+   * @return a {@link Future} to track completion
+   */
+  public Future<Void> setFooter(String footer, int lineCount) {
```

Comment:
Hmm... is `lineCount` actually the line count for `footer`? When I see the bottom of `log()` implementation, I feel like it is the line count of what `mesageLogger` printed plus `footer`? In any case, if `lineCount` is the line count of `footer`, can't we just count that from `footer` instead of accepting it as a param?

Reply:
We could count it from `footer` but I had left that as a parameter since counting it from the footer might be an unnecessary extra computation especially if the footer gets lengthy, but we could leave it out for simplicity of the API.

Reply:
Generally speaking, from the API perspective, removing the unnecessary param eliminates the possibility of misuse, so I think it's better without `lineCount` in the general sense. Even if a footer length is say, 3000, counting from 0 to 3000 even on an old machine should be a blink of an eye. But I also value code simplicity and don't want to introduce more code. I see the tradeoff.

If the caller knows the line count ahead, I wonder if the call can provide `List<String>` from the first place (a `String` element per line). Otherwise, I think I am fine with `lineCounter.`

Reply:
Sounds good, taking the `List<String>` seems like the best way to do it then.